[WIP] Embedded Pressure Field

DamrongGuoy opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/DamrongGuoy/drake/8)
<!-- Reviewable:end -->

DamrongGuoy committed:
[prototype] Embedded Pressure Field on embedding tetrahedral meshes of input surface

DamrongGuoy committed:
Test in hydroelastic-contact tutorial

DamrongGuoy committed:
Read VTK file with signed-distance field for hydroelastics

DamrongGuoy committed:
Clean up, remove unused code, rename ReadVtkPressureValues to ReadVtkToScalarValues()

DamrongGuoy committed:
Add a coarser embedded pressure field (1cm offset, 2cm grid)

DamrongGuoy committed:
Keep only the bell pepper on the table. Distance between surfaces.

DamrongGuoy committed:
"distance" between the 0-level set and the original mesh.

DamrongGuoy committed:
Replay showing the collision mesh enclosing the visual mesh

DamrongGuoy committed:
Auto generate EmPress mesh and signed-distance field.

DamrongGuoy committed:
Refactor code from unit tests into library, so it's usable.

DamrongGuoy committed:
Executable geometry/proximity:empress

DamrongGuoy committed:
Keep only grid-resolution parameter.  No out_offset, no in-offset parameters.

DamrongGuoy committed:
Introduce CalcRMSErrorOfSDField()

DamrongGuoy committed:
Start SDFieldOptimizer

DamrongGuoy committed:
Refactor into optimize_sdfield library

DamrongGuoy committed:
Laplacian smoothing both boundary and interior.

DamrongGuoy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 30 files reviewed, 1 unresolved discussion

___
*[a discussion](https://reviewable.io/reviews/damrongguoy/drake/8#-OLikgbI-QjnG7S9b0D0:-OLikgbJ3SoThFiRMDgv:b-rid3kv) (no related file):*
@amcastro-tri


<!-- Sent from Reviewable.io -->

DamrongGuoy committed:
Write an intermediate file from each iteration of vertex optimization.

DamrongGuoy committed:
Remove data files with no more use

amcastro-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 30 files reviewed, 8 unresolved discussions (waiting on @DamrongGuoy)

___
*[a discussion](https://reviewable.io/reviews/damrongguoy/drake/8#-OLikgbI-QjnG7S9b0D0:-OLkWu9sEy0gERI_qtY7:b-72ruo7) (no related file):*
<details><summary><i>Previously, DamrongGuoy (Damrong Guoy) wrote…</i></summary><blockquote>

@amcastro-tri
</blockquote></details>

Great progress! keep it up
___
*[`geometry/proximity/optimize_sdfield.cc` line 42 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVMDU6INEX6IYbNax:-OLkVMDVEi2mbQ_cQRhZ:blm8tng) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L42)):*
> ```cpp
>   for (int i = 0; i < std::ssize(boundary_to_volume_); ++i) {
>     int bv = boundary_to_volume_.at(i);
>     Vector3d target = CalcLaplacianNewPosition(bv);
> ```

This resembles a Gauss-Seidel approach where we use the latest computed state, and thus the update depends on the order we iterate positions. It can go unstable. 
I think for our purposes a Jacobi update (compute targets using only previous iteration positions, a copy) would be more stable. 
___
*[`geometry/proximity/optimize_sdfield.cc` line 57 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVkhxALItyfHbUPpq:-OLkVkhxALItyfHbUPpr:b-v3ltky) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L57)):*
> ```cpp
>         original_boundary_.tri_bvh(),
>         std::get<FeatureNormalSet>(original_boundary_.feature_normal()));
>     // x^{k+1} = x^{k} - beta * (phi - d) * normal, beta is dimensionless
> ```

here try the projection idea we discussed in Slack. 
Probably provide option with a flag so that we can play with both. 
It should be a simple one-liner change.
___
*[`geometry/proximity/optimize_sdfield.cc` line 60 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVgNRDQlGxdll_Yw5:-OLkVgNRDQlGxdll_Yw6:b-m9gt86) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L60)):*
> ```cpp
>     vertices_.at(bv) =
>         vertices_.at(bv) -
>         beta * (d.signed_distance - target_boundary_distance) * d.gradient;
> ```

always normalize gradient to get a normal.
___
*[`geometry/proximity/optimize_sdfield.cc` line 64 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWTRqANJiciSlj3e9:-OLkWTRr9dOFt1vAn5vu:b-gfcs14) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L64)):*
> ```cpp
> }
> 
> void SDFieldOptimizer::LaplacianInterior(const double alpha) {
> ```

nit,
here and elsewhere. Kinda odd to reuse the same "f" for SDF and Field.



_Suggestion:_
```cpp
SdfFieldOptimizer
```
___
*[`geometry/proximity/optimize_sdfield.cc` line 107 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWD12COCJI1RNzYpx:-OLkWD12COCJI1RNzYpy:b-ljhgh) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L107)):*
> ```cpp
>   const double alpha = 0.3;
>   const double beta = 0.3;
>   const double target_boundary_distance = 1e-3;  // 1mm tolerance.
> ```

consider passing these as a struct to the optimizer so that we can quickly play with values from the main app.
___
*[`geometry/proximity/optimize_sdfield.cc` line 113 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWN0Z-aP8dXTi1EQW:-OLkWN0Z-aP8dXTi1EQX:b-yu1xpe) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L113)):*
> ```cpp
>     // to move the boundary vertices into the interior of the original
>     // surface. Next step, we will move them back outside the original surface.
>     LaplacianBoundary(beta / 10);
> ```

consider making a distinct parameter, say `alpha_exterior`?
___
*[`geometry/proximity/optimize_sdfield.cc` line 133 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWiJo8vkzlXerNFBk:-OLkWiJp3BWXoIeGf9mW:bvhg70p) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L133)):*
> ```cpp
>                                     "Optimized EmbeddedSignedDistanceField");
> 
>     if (std::abs(rms_error - previous_rms_error) < 1e-6) {
> ```

consider relative tolerance, since the erorrs are already in the submilimiter order.



_Suggestion:_
```cpp
std::abs(rms_error - previous_rms_error)/rms_error
```


<!-- Sent from Reviewable.io -->

DamrongGuoy committed:
Improve signed distance fields

DamrongGuoy reviewed (commented):
> Reviewed 1 of 30 files at r1.
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 30 files reviewed, 8 unresolved discussions (waiting on @amcastro-tri)

___
*[a discussion](https://reviewable.io/reviews/damrongguoy/drake/8#-OLikgbI-QjnG7S9b0D0:-OLlTrjW1aB2dsXBJ-Ow:by3k53e) (no related file):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Great progress! keep it up
</blockquote></details>

Thank you for your kind words.  

After trying a few more examples, I was lucky that the bell pepper was close to being a sphere.  

I hope something like octrees will help me out with non-trivial cases. I'd like the octrees to be the next target and pause my feeble vertex smoothing for now if you would kindly approve it.
___
*[`geometry/proximity/optimize_sdfield.cc` line 42 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVMDU6INEX6IYbNax:-OLl4TGz-U0WOn61JKhV:b-8mgzf9) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L42)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This resembles a Gauss-Seidel approach where we use the latest computed state, and thus the update depends on the order we iterate positions. It can go unstable. 
I think for our purposes a Jacobi update (compute targets using only previous iteration positions, a copy) would be more stable.
</blockquote></details>

I refactored and documented it to make it more transparent that it was Jacobi's update.  The update depends on the "copy" of vertices_ in evolving_volume_mesh_, which doesn't change inside this function.
___
*[`geometry/proximity/optimize_sdfield.cc` line 57 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVkhxALItyfHbUPpq:-OLl8ZUaEAnSMc4vExdy:b-dmk3fp) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L57)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

here try the projection idea we discussed in Slack. 
Probably provide option with a flag so that we can play with both. 
It should be a simple one-liner change.
</blockquote></details>

I am sorry that I got stuck on this idea when writing a code. I'll check with you tomorrow.
___
*[`geometry/proximity/optimize_sdfield.cc` line 60 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVgNRDQlGxdll_Yw5:-OLkqF7q7FJZPFCKfIrZ:boyexbg) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L60)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

always normalize gradient to get a normal.
</blockquote></details>

Done. Thanks.
___
*[`geometry/proximity/optimize_sdfield.cc` line 64 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWTRqANJiciSlj3e9:-OLkt-Af4XfTkoUokyH9:buuvf4f) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L64)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

nit,
here and elsewhere. Kinda odd to reuse the same "f" for SDF and Field.
</blockquote></details>

I wasn't clear what it standed for.  I meant it to be "Signed Distance Field Optimizer." 
___
*[`geometry/proximity/optimize_sdfield.cc` line 107 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWD12COCJI1RNzYpx:-OLkp_ytC9hp4jSXBpys:boyexbg) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L107)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

consider passing these as a struct to the optimizer so that we can quickly play with values from the main app.
</blockquote></details>

Done. Thanks.
___
*[`geometry/proximity/optimize_sdfield.cc` line 113 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWN0Z-aP8dXTi1EQW:-OLkn8Ok6ICHJvhdAXDq:b-wyu0zq) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L113)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

consider making a distinct parameter, say `alpha_exterior`?
</blockquote></details>

Done.  It's `alpha_exterior` now. Thanks.
___
*[`geometry/proximity/optimize_sdfield.cc` line 133 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWiJo8vkzlXerNFBk:-OLkmSbO50S17x--yzdp:boyexbg) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L133)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

consider relative tolerance, since the erorrs are already in the submilimiter order.
</blockquote></details>

Done. Thanks.


<!-- Sent from Reviewable.io -->

DamrongGuoy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 30 files reviewed, 8 unresolved discussions (waiting on @amcastro-tri)

___
*[`geometry/proximity/optimize_sdfield.cc` line 64 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWTRqANJiciSlj3e9:-OLlVUfm-1_ziyC9YKz2:b4s19qm) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L64)):*
<details><summary><i>Previously, DamrongGuoy (Damrong Guoy) wrote…</i></summary><blockquote>

I wasn't clear what it standed for.  I meant it to be "Signed Distance Field Optimizer."
</blockquote></details>

Or would you think `SDFOptimizer` (Signed Distance Field Optimizer) would sound better?


<!-- Sent from Reviewable.io -->

amcastro-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 30 files reviewed, 5 unresolved discussions (waiting on @DamrongGuoy)

___
*[a discussion](https://reviewable.io/reviews/damrongguoy/drake/8#-OLikgbI-QjnG7S9b0D0:-OLnhfs39VJoIJGxxLpg:bran4p6) (no related file):*
<details><summary><i>Previously, DamrongGuoy (Damrong Guoy) wrote…</i></summary><blockquote>

Thank you for your kind words.  

After trying a few more examples, I was lucky that the bell pepper was close to being a sphere.  

I hope something like octrees will help me out with non-trivial cases. I'd like the octrees to be the next target and pause my feeble vertex smoothing for now if you would kindly approve it.
</blockquote></details>

Agree. Lets get the few discussions below resolved and merge  so that we  can move forward.
___
*[`geometry/proximity/optimize_sdfield.cc` line 57 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkVkhxALItyfHbUPpq:-OLnggl14Ie_UQ406SqI:bqfrp06) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L57)):*
<details><summary><i>Previously, DamrongGuoy (Damrong Guoy) wrote…</i></summary><blockquote>

I am sorry that I got stuck on this idea when writing a code. I'll check with you tomorrow.
</blockquote></details>

Okay. Anything in particular is blocking you here?
___
*[`geometry/proximity/optimize_sdfield.cc` line 64 at r2](https://reviewable.io/reviews/damrongguoy/drake/8#-OLkWTRqANJiciSlj3e9:-OLngvHK8tfrI6xumIVe:b-6nybcw) ([raw file](https://github.com/damrongguoy/drake/blob/ed3e9b02e7a05c18ac3846f951459cde5690d254/geometry/proximity/optimize_sdfield.cc#L64)):*
<details><summary><i>Previously, DamrongGuoy (Damrong Guoy) wrote…</i></summary><blockquote>

Or would you think `SDFOptimizer` (Signed Distance Field Optimizer) would sound better?
</blockquote></details>

oh, right, it's a mix right now. I think `SdfOptimizer` if we follow our GSG is best. 
That is, when talking you say "SDF" or "Signed Distance Field", not a mix of those.
___
*[`geometry/proximity/optimize_sdfield.cc` line 153 at r3](https://reviewable.io/reviews/damrongguoy/drake/8#-OLnhL3rCF7rE_nX0N-J:-OLnhL3rCF7rE_nX0N-K:bwerrq4) ([raw file](https://github.com/damrongguoy/drake/blob/ba3326311666dbd18ef6afff589f280b53636ed2/geometry/proximity/optimize_sdfield.cc#L153)):*
> ```cpp
>     const double relative_change =
>         std::abs(rms_error - previous_rms_error) / previous_rms_error;
>     if (relative_change < 1e-3) {
> ```

consider moving this hardcoded tolerance and the maximum number of iterations above into the params struct as well. 
That way we have all parameters in one place and we don't need to go fishing hardcoded values around. This is a one-week old code, imainge when it gets older.


<!-- Sent from Reviewable.io -->

amcastro-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 30 files reviewed, 6 unresolved discussions (waiting on @DamrongGuoy)

___
*[`geometry/proximity/make_empress_field.h` line 66 at r3](https://reviewable.io/reviews/damrongguoy/drake/8#-OLni1vr5ALRcNGIo9X-:-OLni1vsA3_ikInmgkgi:b-7nfefb) ([raw file](https://github.com/damrongguoy/drake/blob/ba3326311666dbd18ef6afff589f280b53636ed2/geometry/proximity/make_empress_field.h#L66)):*
> ```objc
> std::pair<std::unique_ptr<VolumeMesh<double>>,
>           std::unique_ptr<VolumeMeshFieldLinear<double, double>>>
> MakeEmPressSDField(const TriangleSurfaceMesh<double>& mesh_M,
> ```

minor, more funny names here. Consider something like `MakeEmbeddedSdf()`. 
SDF is a term of art, so it makes sense to use its acronym. 
However "Em" meanins nothing 

ditto for the file name. Consider `make_embedded_sdf.h` (and cc)


<!-- Sent from Reviewable.io -->

DamrongGuoy committed:
Coarsen sdf meshes; patch vtkUnstructuredGridQuadricDecimation

DamrongGuoy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/damrongguoy/drake/8)* status: 0 of 31 files reviewed, 7 unresolved discussions (waiting on @amcastro-tri)

___
*[`geometry/proximity/test/embedded_pressure_field_test.cc` line 143 at r4](https://reviewable.io/reviews/damrongguoy/drake/8#-OLqzwMMC9D2Mq01Hpve:-OLqzwMMC9D2Mq01Hpvf:b-kz1tcy) ([raw file](https://github.com/damrongguoy/drake/blob/76bd8f311cbd101865721d70b322d3359e187230/geometry/proximity/test/embedded_pressure_field_test.cc#L143)):*
> ```cpp
> //  error: undefined reference to
> //  'drake_vendor::vtkUnstructuredGridQuadricDecimation::New()' collect2: error:
> //  ld returned 1
> ```

No more link errors.  Done.



_Code quote:_
```cpp
// TODO(DamrongGuoy) Fix the following link errors and activate this test again.
//  I believe they are in "@vtk_internal//:vtkFiltersCore"
//  - vtkCleanPolyData.h
//  - vtkDelaunay3D.h
//  - vtkUnstructuredGridQuadricDecimation.h
//  bazel-out/k8-opt/...::TetrahedralFieldDecimator_SmokeTest_Test::TestBody():
//  error: undefined reference to 'drake_vendor::vtkCleanPolyData::New()'
//  bazel-out/k8-opt/...::TetrahedralFieldDecimator_SmokeTest_Test::TestBody():
//  error: undefined reference to 'drake_vendor::vtkDelaunay3D::New()'
//  bazel-out/k8-opt/...::TetrahedralFieldDecimator_SmokeTest_Test::TestBody():
//  error: undefined reference to
//  'drake_vendor::vtkUnstructuredGridQuadricDecimation::New()' collect2: error:
//  ld returned 1
```


<!-- Sent from Reviewable.io -->

DamrongGuoy committed:
Split into two executables: empress0 and empress1

- empress0 generates a signed distance field and its support mesh from an input surface mesh.
- empress1 improves result from empress0 by mesh smoothing and coarsening.
- Right now coarsen meshes by a conservative fraction = 0.5 to preserve zero-level set.

DamrongGuoy committed:
Preserve 0-level surface by disallowing edge contractions that change sign.

- Disallow edge contraction if the new position has the opposite sign of
  the field variable.  This condition preserves the zero-level set implicit
  surface better.

DamrongGuoy committed:
Adjust for fine meshes.  No more sub-sampling in a tetrahedron.

DamrongGuoy committed:
empress2 for coarsening, patch vtkUnstructuredGridQuadricDecimation.cxx for more edge contractions.
